### PR TITLE
Add robustness benchmarking toolkit

### PR DESCRIPTION
**Pull request by michaelisc**
This pull request contains tools to benchmark the robustness of object detection models to common image corruptions. The corresponding [robust detection benchmark](https://github.com/bethgelab/robust-detection-benchmark) was defined in ["Benchmarking Robustness in Object Detection: Autonomous Driving when Winter is Coming"](https://arxiv.org/abs/1907.07484).

We tried to design the code in a way that it minimally infer with the current workflow, while adding functionality to evaluate any model implemented in mmdetection for robustness. To corrupt and distort the images we uses the [imagecorruptions](https://github.com/bethgelab/imagecorruptions) package we separately developed and distribute through PyPI.

### CODE REVIEW

**Comment by michaelisc** [comment]
Not sure, why the build fails for the last commit. I only moved the readmes, so your original README.md stays in place.

File: ROBUSTNESS_BENCHMARKING.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+```shell
+# single-gpu testing
+python tools/test_c.py ${CONFIG_FILE} ${CHECKPOINT_FILE} [--out ${RESULT_FILE}] [--eval ${EVAL_METRICS}]
+```
+
+Alternatively different group of corruptions can be selected.
+
+```shell
+# noise
+python tools/test_c.py ${CONFIG_FILE} ${CHECKPOINT_FILE} [--out ${RESULT_FILE}] [--eval ${EVAL_METRICS}] --corruptions noise
```
Comment (hellock) [inline ROBUSTNESS_BENCHMARKING.md]: `test_c.py` has been renamed to `test_robustness.py`?

File: mmdet/datasets/custom.py [hunk 2]
```diff
@@ -11,6 +11,8 @@
 from .utils import to_tensor, random_scale
 from .extra_aug import ExtraAugmentation
 
+from imagecorruptions import corrupt
```
Comment (hellock) [inline mmdet/datasets/custom.py]: Third-party imports may be moved before relative imports.

File: mmdet/datasets/custom.py [hunk 3]
```diff
@@ -178,6 +186,10 @@ def prepare_train_img(self, idx):
         img_info = self.img_infos[idx]
         # load image
         img = mmcv.imread(osp.join(self.img_prefix, img_info['filename']))
+        # corruption
+        if self.corruption is not None:
+            img = corrupt(img, severity=self.corruption_severity,
+                          corruption_name=self.corruption)
```
Comment (hellock) [inline mmdet/datasets/custom.py]: These lines need to be formatted by yapf with the specified [style configurations](https://github.com/open-mmlab/mmdetection/blob/master/.style.yapf).

File: tools/robustness_eval.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+    num_distortions = len(list(eval_output.keys()))
+    results = np.zeros((num_distortions, 6, 20), dtype='float32')
+
+    for i, distortion in enumerate(eval_output):
+        for severity in eval_output[distortion]:
+            mAP = [eval_output[distortion][severity][j]['ap']
+                   for j in range(len(eval_output[distortion][severity]))]
+            results[i, severity, :] = mAP
+            # if verbose > 0:
+            #     print(distortion, severity, mAP)
```
Comment (hellock) [inline tools/robustness_eval.py]: These comments may be removed.

File: tools/robustness_eval.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+    if aggregate == 'benchmark':
+        mPC = np.mean(results[:15, 1:, :], axis=(0, 1))
+    else:
+        mPC = np.mean(results[:, 1:, :], axis=(0, 1))
+    rPC = mPC/P
+
+    print("\nmodel: {}".format(osp.basename(filename)))
+    if metric is None:
+        if 'P' in prints:
+            print("Performance on Clean Data [P] ({})"
```
Comment (hellock) [inline tools/robustness_eval.py]: To be consistent with other files, single quote is preferred.

File: tools/robustness_eval.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+        metrics = [metric]
+
+    for metric_name in metrics:
+        assert metric_name in [
+            'AP', 'AP50', 'AP75', 'APs', 'APm', 'APl',
+            'AR1', 'AR10', 'AR100', 'ARs', 'ARm', 'ARl'
+        ]
+
+    with open(filename, "rb") as f:
+        eval_output = pickle.load(f)
```
Comment (hellock) [inline tools/robustness_eval.py]: These two lines can be simply replaced with `mmcv.load(filename)`.

File: ROBUSTNESS_BENCHMARKING.md (hunk 1 above)
Comment (michaelisc) [reply ROBUSTNESS_BENCHMARKING.md]: Yes

File: mmdet/datasets/custom.py (hunk 2 above)
Comment (michaelisc) [reply mmdet/datasets/custom.py]: Done

File: mmdet/datasets/custom.py (hunk 3 above)
Comment (michaelisc) [reply mmdet/datasets/custom.py]: Done

File: tools/robustness_eval.py (hunk 4 above)
Comment (michaelisc) [reply tools/robustness_eval.py]: Done

File: tools/robustness_eval.py (hunk 5 above)
Comment (michaelisc) [reply tools/robustness_eval.py]: Done

File: tools/robustness_eval.py (hunk 6 above)
Comment (michaelisc) [reply tools/robustness_eval.py]: Done

File: ROBUSTNESS_BENCHMARKING.md (hunk 1 above)
Comment (michaelisc) [reply ROBUSTNESS_BENCHMARKING.md]: Changed the name

**Comment by hellock** [comment]
Hi there are some conflicts. Could you fix that?

**Comment by hellock** [comment]
Sorry that I noted that there were some `isort` inconsistency. I will fix that.

**Comment by michaelisc** [comment]
Give me a minute. I am currently fixing the yapf errors

**Comment by michaelisc** [comment]
The merge conflicts and yapf errors should be fixed.

**Pull request merged**